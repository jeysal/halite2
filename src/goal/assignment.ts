import GameMap from 'halite/GameMap';
import { OwnShip } from 'halite/Ship';
import { maxBy } from 'ramda';

import Goal from '../goal/goal';

interface ShipGoal<Turn> {
  ship: OwnShip<Turn>;
  goal: Goal<Turn>;
}

const VALUE_IMPACT = 0.5;
const SUITABILITY_IMPACT = 0.5;

const shipGoalValue = <Turn>(
  ship: OwnShip<Turn>,
  goal: Goal<Turn>,
  gm: GameMap<Turn>,
): number =>
  VALUE_IMPACT * goal.value(gm) +
  SUITABILITY_IMPACT * goal.suitability(ship, gm);

const assignGoalsToShips = <Turn>(
  goals: Goal<Turn>[],
  gm: GameMap<Turn>,
): ShipGoal<Turn>[] => {
  const ownShips = gm.myShips;

  const shipGoalValues = ownShips.map(ship =>
    goals.map(goal => ({ ship, goal, val: shipGoalValue(ship, goal, gm) })),
  );

  const goalsWithShips = shipGoalValues.reduce((gwss, gVals) => {
    const gValsBelowLimit = gVals.filter(gVal => {
      const entry = gwss.find(gws => gws.goal === gVal.goal);
      return !entry || entry.ships.length < entry.goal.maxShips();
    });
    if (gValsBelowLimit.length) {
      const bestVal = gValsBelowLimit.reduce(maxBy(value => value.val));
      return gwss.map(
        goalWithShips =>
          bestVal && goalWithShips.goal === bestVal.goal
            ? {
                goal: goalWithShips.goal,
                ships: [...goalWithShips.ships, bestVal.ship],
              }
            : goalWithShips,
      );
    }
    return gwss;
  }, goals.map(goal => ({ goal, ships: [] as OwnShip<Turn>[] })));

  // TODO goal minima

  return goalsWithShips.reduce(
    (acc, gws) => [
      ...acc,
      ...gws.ships.map(ship => ({ ship, goal: gws.goal })),
    ],
    [] as ShipGoal<Turn>[],
  );
};

export default assignGoalsToShips;
