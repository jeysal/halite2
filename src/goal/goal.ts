import { Action } from 'halite/Game';
import GameMap from 'halite/GameMap';
import { OwnShip } from 'halite/Ship';

export default interface Goal<Turn = number> {
  value(gm: GameMap<Turn>): number;
  minShips(): number;
  maxShips(): number;
  suitability(ship: OwnShip<Turn>, gm: GameMap<Turn>): number;
  strive(ship: OwnShip<Turn>, gm: GameMap<Turn>): string & Action<Turn>;
};
