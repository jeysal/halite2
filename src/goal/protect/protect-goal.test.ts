import DockingStatus from 'halite/DockingStatus';
import GameMap from 'halite/GameMap';
import Planet, { OwnPlanet } from 'halite/Planet';
import Ship, { OwnShip } from 'halite/Ship';

import { goalStats } from '../../test/goal';
import { emptyMap } from '../../test/util';

import ProtectGoal, { determineProtectGoals } from './protect-goal';

describe('value', () => {
  test('is high if an enemy is close', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlayerShips(0, [{ ownerId: 0, x: 17, y: 17 }]);
    gm.addPlayerShips(1, [{ ownerId: 1, x: 15, y: 15 }]);

    expect(
      new ProtectGoal(new Planet(gm, {
        ownerId: 0,
        x: 20,
        y: 20,
        radius: 3,
        dockedShipIds: [100, 101],
      }) as OwnPlanet).value(gm),
    ).toMatchSnapshot();
  });

  test('is low if no enemy is close', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlayerShips(0, [{ ownerId: 0, x: 17, y: 17 }]);
    gm.addPlayerShips(1, [{ ownerId: 1, x: 85, y: 85 }]);

    expect(
      new ProtectGoal(new Planet(gm, {
        ownerId: 0,
        x: 20,
        y: 20,
        radius: 3,
        dockedShipIds: [100, 101],
      }) as OwnPlanet).value(gm),
    ).toMatchSnapshot();
  });

  test('is high if many ships are docked', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlayerShips(0, [{ ownerId: 0, x: 17, y: 17 }]);
    gm.addPlayerShips(1, [{ ownerId: 1, x: 35, y: 35 }]);

    expect(
      new ProtectGoal(new Planet(gm, {
        ownerId: 0,
        x: 20,
        y: 20,
        radius: 3,
        dockedShipIds: [100, 101, 102, 103, 104, 105],
      }) as OwnPlanet).value(gm),
    ).toMatchSnapshot();
  });

  test('is low if just one ship is docked', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlayerShips(0, [{ ownerId: 0, x: 17, y: 17 }]);
    gm.addPlayerShips(1, [{ ownerId: 1, x: 35, y: 35 }]);

    expect(
      new ProtectGoal(new Planet(gm, {
        ownerId: 0,
        x: 20,
        y: 20,
        radius: 3,
        dockedShipIds: [100],
      }) as OwnPlanet).value(gm),
    ).toMatchSnapshot();
  });
});

describe('minShips', () => {
  test('is 1', () => {
    expect(
      new ProtectGoal(new Planet(emptyMap, {
        dockedShipIds: [100],
      }) as OwnPlanet).minShips(),
    ).toBe(1);
  });
});
describe('maxShips', () => {
  test('returns a reasonable maximum for a planet with 1 docked ship', () => {
    expect(
      new ProtectGoal(new Planet(emptyMap, {
        dockedShipIds: [100],
      }) as OwnPlanet).maxShips(),
    ).toMatchSnapshot();
  });

  test('returns a reasonable maximum for a planet with many docked ships', () => {
    expect(
      new ProtectGoal(new Planet(emptyMap, {
        dockedShipIds: [100, 101, 102, 103, 104, 105],
      }) as OwnPlanet).maxShips(),
    ).toMatchSnapshot();
  });
});

describe('suitability', () => {
  test('is high for an undocked nearby ship', () => {
    expect(
      new ProtectGoal(new Planet(emptyMap, {
        x: 15,
        y: 15,
      }) as OwnPlanet).suitability(
        new Ship(emptyMap, 0, {
          x: 20,
          y: 20,
          dockingStatus: DockingStatus.UNDOCKED,
        }) as OwnShip,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });

  test('is low for a docked nearby ship', () => {
    expect(
      new ProtectGoal(new Planet(emptyMap, {
        x: 15,
        y: 15,
      }) as OwnPlanet).suitability(
        new Ship(emptyMap, 0, {
          x: 20,
          y: 20,
          dockingStatus: DockingStatus.DOCKED,
        }) as OwnShip,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });

  test('is low for an undocked distant ship', () => {
    expect(
      new ProtectGoal(new Planet(emptyMap, {
        x: 15,
        y: 15,
      }) as OwnPlanet).suitability(
        new Ship(emptyMap, 0, {
          x: 100,
          y: 100,
          dockingStatus: DockingStatus.UNDOCKED,
        }) as OwnShip,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });
});

describe('strive', () => {
  test('navigates between the planet and the closest enemy ship', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlayerShips(1, [{ x: 20, y: 20 }]);

    expect(
      new ProtectGoal(new Planet(gm, { x: 10, y: 10 }) as OwnPlanet).strive(
        new Ship(gm, 0, {
          id: 100,
          x: 10,
          y: 20,
          dockingStatus: DockingStatus.UNDOCKED,
        }) as OwnShip,
        gm,
      ),
    ).toMatchSnapshot();
  });

  test('undocks a docked ship', () => {
    expect(
      new ProtectGoal(new Planet(emptyMap, {}) as OwnPlanet).strive(
        new Ship(emptyMap, 0, {
          id: 100,
          dockingStatus: DockingStatus.DOCKED,
        }) as OwnShip,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });
});

describe('determineProtectGoals', () => {
  test('creates a protect goal for each own planet', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlanets([
      {
        id: 1000,
        ownerId: 0,
        x: 10,
        y: 10,
        radius: 3,
        dockedShipIds: [100, 101],
      },
    ]);
    gm.addPlayerShips(0, [
      {
        id: 100,
        x: 15,
        y: 15,
        radius: 1,
        dockingStatus: DockingStatus.DOCKED,
        dockedPlanetId: 1000,
      },
      {
        id: 101,
        x: 5,
        y: 15,
        radius: 1,
        dockingStatus: DockingStatus.DOCKED,
        dockedPlanetId: 1000,
      },
    ]);
    gm.addPlayerShips(1, [
      {
        id: 200,
        x: 20,
        y: 20,
        radius: 1,
        dockingStatus: DockingStatus.UNDOCKED,
      },
    ]);

    expect(
      determineProtectGoals(gm).map(goal =>
        goalStats(
          goal,
          new Ship(gm, 0, {
            id: 110,
            x: 10,
            y: 20,
            radius: 1,
            dockingStatus: DockingStatus.UNDOCKED,
          }),
          gm,
        ),
      ),
    ).toMatchSnapshot();
  });
});
