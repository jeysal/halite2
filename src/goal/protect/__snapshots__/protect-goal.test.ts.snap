// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`determineProtectGoals creates a protect goal for each own planet 1`] = `
Array [
  Object {
    "action": "t 110 6 325",
    "maxShips": 6,
    "minShips": 1,
    "suitability": 0.9646446609406727,
    "value": 0.5098155246886393,
  },
]
`;

exports[`maxShips returns a reasonable maximum for a planet with 1 docked ship 1`] = `3`;

exports[`maxShips returns a reasonable maximum for a planet with many docked ships 1`] = `18`;

exports[`strive navigates between the planet and the closest enemy ship 1`] = `"t 100 6 315"`;

exports[`strive undocks a docked ship 1`] = `"u 100"`;

exports[`suitability is high for an undocked nearby ship 1`] = `0.975`;

exports[`suitability is low for a docked nearby ship 1`] = `0.475`;

exports[`suitability is low for an undocked distant ship 1`] = `0.575`;

exports[`value is high if an enemy is close 1`] = `0.5395922388920072`;

exports[`value is high if many ships are docked 1`] = `0.5919801340635383`;

exports[`value is low if just one ship is docked 1`] = `0.29599006703176917`;

exports[`value is low if no enemy is close 1`] = `0.00750628127905511`;
